File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderSubAccount.java
package uk.gov.justice.hmpps.prison.repository.jpa.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.IdClass;
import jakarta.persistence.Table;
import java.io.Serializable;
import java.math.BigDecimal;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@IdClass(OffenderSubAccount.Pk.class)
@Table(name = "OFFENDER_SUB_ACCOUNTS")
public class OffenderSubAccount {

    @Id
    @Column(name = "CASELOAD_ID", nullable = false, insertable = false, updatable = false)
    private String prisonId;

    @Id
    @Column(name = "OFFENDER_ID", nullable = false, insertable = false, updatable = false)
    private Long offenderId;

    @Id
    @Column(name = "TRUST_ACCOUNT_CODE", nullable = false, insertable = false, updatable = false)
    private Long accountCode;

    @Column(name = "BALANCE", nullable = false)
    private BigDecimal balance;

    @EqualsAndHashCode
    @NoArgsConstructor
    @AllArgsConstructor
    @ToString
    public static class Pk implements Serializable {
        private String prisonId;
        private Long offenderId;
        private Long accountCode;
    }
}

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderSubAccount.kt
package uk.gov.justice.hmpps.prison.repository.jpa.model

import jakarta.persistence.Column
import jakarta.persistence.Embeddable
import jakarta.persistence.EmbeddedId
import jakarta.persistence.Entity
import jakarta.persistence.Table
import org.hibernate.Hibernate
import uk.gov.justice.hmpps.prison.repository.jpa.helper.EntityOpen
import java.io.Serializable
import java.math.BigDecimal

@Embeddable
data class OffenderSubAccountId(
  @Column(name = "CASELOAD_ID", insertable = false)
  val prisonId: String,

  @Column(name = "OFFENDER_ID", nullable = false)
  val offenderId: Long,

  @Column(name = "TRUST_ACCOUNT_CODE", nullable = false)
  val accountCode: Long,
) : Serializable

@Entity
@Table(name = "OFFENDER_SUB_ACCOUNTS")
@EntityOpen
class OffenderSubAccount(

  @EmbeddedId
  val id: OffenderSubAccountId,

  @Column(name = "BALANCE", nullable = false)
  var balance: BigDecimal,

  @Column(name = "HOLD_BALANCE")
  var holdBalance: BigDecimal? = null,
) {
  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || Hibernate.getClass(this) != Hibernate.getClass(other)) return false
    other as OffenderSubAccount
    return id == other.id
  }

  override fun hashCode(): Int = id.hashCode()
}
